MAINT: Use `spin lint` in CI

ganesh-k13 opened the pull request:
### Changes
- use `spin lint` for CI
- ~~use spin notes for CI~~

### Testing

- Added a tmp [commit](https://github.com/numpy/numpy/pull/25103/commits/17ee666095a630605821962ffd5b73c260b0b64e) with linter errors: [failure](https://github.com/numpy/numpy/actions/runs/6859950172/job/18653011873?pr=25103#step:5:1)

### Notes
- related
  - #24080
  - https://github.com/numpy/numpy/pull/24983
  - ~~https://github.com/numpy/numpy/pull/25017~~

ganesh-k13 committed:
ENH: Use `spin` commands for CI

- use `spin lint` for CI
- use `spin notes` for CI

ganesh-k13 reviewed (commented):
Need to add `spin` as a lint requirement now. Would it make sense to merge `linter_requirements.txt` with `build_requirements.txt` now? Since lint checks should technically be part of the development workflow?

mattip commented:
> Would it make sense to merge linter_requirements.txt with build_requirements.txt now? Since lint checks should technically be part of the development workflow?

I would personally prefer they remain separate, but am willing to be convinced that they can be merged.

rgommers commented:
+1 for keeping separate, `build_requirements` is for building and is used a fair amount. linting is only done in 1-2 jobs. no need to install unnecessary linting tools in regular CI jobs.

ganesh-k13 commented:
Thanks for the inputs folks, I have modified the requirements to include spin.

Testing:
- Added a tmp [commit](https://github.com/numpy/numpy/pull/25103/commits/17ee666095a630605821962ffd5b73c260b0b64e) linter errors: [failure](https://github.com/numpy/numpy/actions/runs/6859950172/job/18653011873?pr=25103#step:5:1)
- Release notes: [Will update when done, taking ages]

seberg commented on linter_requirements.txt:
```diff
@@ -1,2 +1,3 @@
 pycodestyle==2.8.0
 GitPython>=3.1.30
+spin==0.7
```
Ci failure seems real, I guess the `spin lint` is trying to rebuild so it needs a pretty full dev env?
Also, should this be `>=`?

ganesh-k13 commented on numpy/conftest.py:
```diff
@@ -32,9 +32,7 @@
-hypothesis.configuration.set_hypothesis_home_dir(
-    os.path.join(tempfile.gettempdir(), ".hypothesis")
-)
+hypothesis.configuration.set_hypothesis_home_dir(os.path.join(tempfile.gettempdir(), ".hypothesis"))
```
It works:
```
Run spin lint
  spin lint
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.12.12/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.12.12/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.12/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.12/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.12/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.12.12/x64/lib
    BASE_REF: main
Running Ruff Check...
E501 Line too long (100 > 88)
  --> numpy/conftest.py:35:89
   |
33 | # Use a known and persistent tmpdir for hypothesis' caches, which
34 | # can be automatically cleared by the OS or user.
35 | hypothesis.configuration.set_hypothesis_home_dir(os.path.join(tempfile.gettempdir(), ".hypothesis"))
   |                                                                                         ^^^^^^^^^^^^
36 |
37 | # We register two custom profiles for Numpy - for details see
   |
Found 1 error.
Error: Process completed with exit code 1.
```

ganesh-k13 replied:
https://github.com/numpy/numpy/actions/runs/21509690315/job/61973451584?pr=25103

ganesh-k13 committed:
MAINT: Removed ``spin notes`` as ``docs`` does it
